### DISCUSSION TITLE
Naked LGT8F328 crashes when writing to EEPROM

### DISCUSSION THREAD

**Discussion by alxgarza** [dbuezas/lgt8fx#288] [Q&A] [answered]
I am trying to program a LGT8F328 with bare minimum code. No Arduino libraries.
I got everything working except EEPROM write.
Whenever I write to EEPROM the LGT8F328 hangs.

This is the function I am using to write one byte to EEPROM:

void DrvEEPROM_Init(void)
{
    ECCR = (1 << EWEN); // 0x80
    ECCR = (1 << EEN) | (1 << CP1) | (1 << CP0); //0x4C  
}

void DrvEEPROM_ProgByte(uint16_t u16Addr, uint8_t u8Data){
    uint8_t __bk_sreg = SREG;
    
    cli();
    
    /* Wait for completion of previous write */
    while(EECR & (1<<EEPE))
    
    // address
    EEARH = u16Addr >> 8;
    EEARL = (u16Addr) & 0xff;
    
    // data
    EEDR = u8Data;
     
    /* Write logical one to EEMPE */
    EECR2 = (1<<EEMPE); // 0x04
    
    /* Start eeprom write by setting EEPE */
    EECR2 = (1<<EEPE);  // 0x02        CODE STOPS EXECUTION HERE!!!!!!!!!!
    
    asm("nop");
    asm("nop");
    
    SREG = __bk_sreg;
}

Any idea what I am missing?

**Comment by alxgarza** [comment]
Forgot to comment,
EECR2 = EECR (They have the same definition)

**Reply by vklimk** [reply]
Ensure that system clock required for eeprom operations is configured and started.

**Reply by alxgarza** [reply]
You nailed it!
You need to have the internal RC clock enabled for EEPROM to work.  Since I use an external crystal, I did not have the internal clock enabled.

Thanks.

**Comment by LaZsolt** [comment, marked as answer]
LGT8F328P ha no EEPROM. You could store data in a small part of the program flash with hardware / software emulated EEPROM functions.
You could rewrite those bytes which was empty (0xFF) before. If you want to rewrite the same byte you need to clear 1 KB program flash first.
This procedure is so complex that it is recommended to use the LGT EEPROM library.
